File: trapezoid.exs
defmodule Trapezoid do

  # Mathematical function
  def fx(x) do
    :math.sin(x) * :math.cos(x) * :math.sqrt(x)
  end

  # --- Worker Logic ---
  def worker(master_pid, left, step, intervals) do
    # Calculate area for this chunk
    area = calculate_chunk(left, step, intervals, 0.0)
    send(master_pid, {:result, area})
  end

  # Tail-recursive math loop
  defp calculate_chunk(_x, _step, 0, acc_area), do: acc_area
  defp calculate_chunk(x, step, count, acc_area) do
    # Trapezoidal rule math
    trap_area = (fx(x) + fx(x + step)) / 2.0 * step
    calculate_chunk(x + step, step, count - 1, acc_area + trap_area)
  end

  # --- Master Logic ---
  def master(total_workers, workers_finished, total_area, start_time) do
    if workers_finished == total_workers do
      end_time = System.system_time(:millisecond)
      IO.puts("LOG_END:#{end_time}")
      IO.puts("Result Area: #{total_area}")
      IO.puts("Time taken: #{end_time - start_time} ms")
      System.halt(0)
    else
      receive do
        {:result, area} ->
          master(total_workers, workers_finished + 1, total_area + area, start_time)
      end
    end
  end

  # --- Main ---
  def run do
    # 100 Million intervals for heavy CPU load
    total_intervals = 100_000_000
    num_workers = 100
    intervals_per_worker = div(total_intervals, num_workers)

    left_boundary = 1.0
    right_boundary = 100.0
    step = (right_boundary - left_boundary) / total_intervals

    IO.puts("--- STARTING TRAPEZOID (Elixir) ---")
    start_time = System.system_time(:millisecond)
    IO.puts("LOG_START:#{start_time}")

    # Spawn Master
    master_pid = spawn(fn -> master(num_workers, 0, 0.0, start_time) end)

    # Spawn Workers
    Enum.each(0..(num_workers - 1), fn i ->
      w_left = left_boundary + (i * intervals_per_worker * step)
      spawn(fn -> worker(master_pid, w_left, step, intervals_per_worker) end)
    end)

    Process.sleep(:infinity)
  end
end

Trapezoid.run()

File: Trapezoid.java
import akka.actor.*;

public class Trapezoid {

    // Mathematical function to integrate
    static double fx(double x) {
        return Math.sin(x) * Math.cos(x) * Math.sqrt(x);
    }

    // --- Messages ---
    static class WorkMessage {
        public final double left, right, step;
        public final long intervals;
        public WorkMessage(double left, double right, double step, long intervals) {
            this.left = left; this.right = right; this.step = step; this.intervals = intervals;
        }
    }
    
    static class ResultMessage {
        public final double area;
        public ResultMessage(double area) { this.area = area; }
    }

    // --- Worker Actor ---
    static class Worker extends AbstractActor {
        @Override
        public Receive createReceive() {
            return receiveBuilder()
                .match(WorkMessage.class, msg -> {
                    double area = 0.0;
                    double x = msg.left;
                    for (long i = 0; i < msg.intervals; i++) {
                        // Trapezoidal rule math
                        area += (fx(x) + fx(x + msg.step)) / 2.0 * msg.step;
                        x += msg.step;
                    }
                    getSender().tell(new ResultMessage(area), getSelf());
                }).build();
        }
    }

    // --- Master Actor ---
    static class Master extends AbstractActor {
        private final int totalWorkers;
        private int workersFinished = 0;
        private double totalArea = 0.0;
        private final long startTime;

        public Master(int totalWorkers, long startTime) {
            this.totalWorkers = totalWorkers;
            this.startTime = startTime;
        }

        @Override
        public Receive createReceive() {
            return receiveBuilder()
                .match(ResultMessage.class, msg -> {
                    totalArea += msg.area;
                    workersFinished++;
                    
                    if (workersFinished == totalWorkers) {
                        long end = System.currentTimeMillis();
                        System.out.println("LOG_END:" + end);
                        System.out.println("Result Area: " + totalArea);
                        System.out.println("Time taken: " + (end - startTime) + " ms");
                        getContext().getSystem().terminate();
                    }
                }).build();
        }
    }

    // --- Main ---
    public static void main(String[] args) {
        // Total intervals to calculate (100 Million for heavy CPU load)
        long totalIntervals = 2000000000L;
        int numWorkers = 100;
        long intervalsPerWorker = totalIntervals / numWorkers;
        
        double leftBoundary = 1.0;
        double rightBoundary = 100.0;
        double step = (rightBoundary - leftBoundary) / totalIntervals;

        ActorSystem system = ActorSystem.create("TrapezoidSystem");
        
        System.out.println("--- STARTING TRAPEZOID (Java) ---");
        long start = System.currentTimeMillis();
        System.out.println("LOG_START:" + start);

        ActorRef master = system.actorOf(Props.create(Master.class, numWorkers, start));

        // Create workers and distribute the math
        for (int i = 0; i < numWorkers; i++) {
            ActorRef worker = system.actorOf(Props.create(Worker.class));
            double wLeft = leftBoundary + (i * intervalsPerWorker * step);
            double wRight = wLeft + (intervalsPerWorker * step);
            
            worker.tell(new WorkMessage(wLeft, wRight, step, intervalsPerWorker), master);
        }
    }
}